File: stringMatch/PermutationString567/PermutationString567.go
/*
567. 字符串的排列
给定两个字符串 s1 和 s2，写一个函数来判断 s2 是否包含 s1 的排列。

换句话说，第一个字符串的排列之一是第二个字符串的子串。
输入: s1 = "ab" s2 = "eidbaooo"
输出: True
解释: s2 包含 s1 的排列之一 ("ba").
*/
package PermutationString567

func checkInclusion(s1 string, s2 string) bool {
	left := 0
	right := 0
	needs := make(map[byte]int)
	windows := make(map[byte]int)
	isShrink := 0
	for _, c := range s1 {
		needs[byte(c)]++
	}
	for right < len(s2) {
		c := s2[right]
		right++
		if _, ok := needs[c]; ok {
			windows[c]++
			if needs[c] == windows[c] {
				isShrink++
			}
		}
		for right-left >= len(s1) {
			if isShrink == len(needs) {
				return true
			}
			d := s2[left]
			left++
			if _, ok := needs[d]; ok {
				if needs[d] == windows[d] {
					isShrink--
				}
				windows[d]--
			}
		}
	}
	return false
}

File: stringMatch/PermutationString567/PermutationString567.cc
/**
 * 567. Permutation in String
 * Given two strings s1 and s2, write a function to return true if s2 contains the permutation of s1.
 * In other words, one of the first string's permutations is the substring of the second string.
 * 
 * example 
 * Input: s1 = "ab" s2 = "eidbaooo"
 * Output: True
 * Explanation: s2 contains one permutation of s1 ("ba").
 * */
#include<string>
#include<iostream>
#include<unordered_map>
using namespace std;
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        unordered_map<char, int> needs,windows;
        for(char c:s1) needs[c]++;
        int left = 0, right = 0;
        int valid = 0;
        while (right < s2.size()){
            char c = s2[right];
            right ++;
            if (needs.count(c)){
                windows[c]++;
                if (windows[c] ==  needs[c])
                    valid++; 
            }
            //当窗口大小大于 s1.size 时 缩小窗口
            while (right - left >= s1.size()){
                //合法的排列
                if (valid == needs.size()) return true;
                char d = s2[left];
                left ++;
                if(needs.count(d)){
                    if(windows[d] == needs[d])
                        valid --;
                    windows[d]--;
                }
            }
        }
        return false;        
    }
};